src/compiler/checker.ts

```diff
@@ -5437,11 +5517,32 @@ module ts {
 
                 return type;
             }
+            
+            function narrowTypeByCallExpression(type: Type, expr: CallExpression, assumeTrue: boolean): Type {
+                // Check that type is not any, assumed result is true.
+                if (type.flags & TypeFlags.Any || !assumeTrue) {
+                    return type;
+                }
+                let callType = checkCallExpression(expr);
+                if (callType.flags & TypeFlags.TypeGuard) {
+                    if (expr.arguments && expr.arguments[(<TypeGuardType>callType).parameterIndex]) {
+                        if (getSymbolAtLocation(expr.arguments[(<TypeGuardType>callType).parameterIndex]) === symbol) {
+                            let narrowedType = (<TypeGuardType>callType).type;
+                            if (isTypeAssignableTo(narrowedType, type)) {
+                               return narrowedType;

```

Comment:
you would also want to handle the case of this. or make it an error in the parser for now. i think we can enable this in a different change.